File: leetcode/algorithms/278_first_bad_version/main.js
/**
 * Definition for isBadVersion()
 *
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

isBadVersion = function (version) {
  return false;
};

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function (isBadVersion) {
  /**
   * @param {integer} n Total versions
   * @return {integer} The first bad version
   */
  return function (n) {
    let left = 0;
    let right = n;

    while (left < right) {
      const mid = Math.floor(left + (right - left) / 2);

      if (isBadVersion(mid)) {
        right = mid;
      } else {
        left = mid + 1;
      }
    }

    return left;
  };
};


// Best Solution
var bestSolution = function (isBadVersion) {
  /**
   * @param {integer} n Total versions
   * @return {integer} The first bad version
   */
  return function (n) {
    // binary search
    var start = 1,
      end = n;
    while (start < end) {
      var mid = Math.floor(start + (end - start) / 2);
      if (isBadVersion(mid)) {
        end = mid; // look on left side of mid
      } else {
        start = mid + 1; // look on the right side of mid
      }
    }
    return start;
  };
};

File: leetcode/algorithms/278_first_bad_version/main.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */
class VersionControl {
    boolean isBadVersion(int version) {
        return false;
    }
}

class FirstBadVersion extends VersionControl {
    public int firstBadVersion(int n) {
        int left = 0;
        int right = n;

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (isBadVersion(mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }


    // Solution
    // Solution 1: First Bad Version Left
    public int solution1(int n) {
        int i = 1;
        int j = n;

        while (i < j) {
            int mid = i + (j - i) / 2;

            if (isBadVersion(mid)) {
                j = mid;
            } else {
                i = mid + 1;
            }
        }

        return i;
    }

    // Solution 2: First Bad Version Right
    public int solution2(int n) {
        if (isBadVersion(1)) {
            return 1;
        }

        int i = 1;
        int j = n;

        while (i < j) {
            int mid = i + (j - i) / 2 + 1;
            if (!isBadVersion(mid)) {
                i = mid;
            } else {
                j = mid - 1;
            }
        }

        return j + 1;
    }
}
